add-channel-single.py

```diff
@@ -0,0 +1,202 @@
+    """Print the usage for the script."""
+    print("{} <charm-name> <channel>".format(sys.argv[0]))
+    print("")
+    print("Change or add the juju channel to the bundles for the charm.")
+
+
+def find_bundles_dirs(charm_dir: str) -> List[str]:
+    """Find the directory with bundles.
+
+    :param charm_dir: the root dir of the charm
+    :returns: the dir where the bundles are.
+    """
+    paths = (('tests', 'bundles'),
+             ('tests', 'bundles', 'overlays'),
+             ('src', 'tests', 'bundles'),
+             ('src', 'tests', 'bundles', 'overlays'))
+
+    found_list = []
+    for path in paths:
+        print(path)
+        dir_ = os.path.abspath(os.path.join(charm_dir, *path))
+        if os.path.isdir(dir_):
+            found_list.append(dir_)
+    return found_list
+
+
+def find_bundles(bundles_dir: str) -> List[str]:
+    """Get a list of all bundles, including any overlays.
+
+    :param bundles_dir: the directory with bundles (hopefully).
+    :returns: List of filenames of bundles (yaml or yaml.j2)
+    """
+    print("scanning:", bundles_dir)
+    bundles = []
+    for file in os.listdir(bundles_dir):
+        path = os.path.join(bundles_dir, file)
+        if os.path.isfile(path):
+            if path.endswith(".yaml") or path.endswith(".yaml.j2"):
+                bundles.append(path)
+    return bundles
+
+
+def find_bundles_in_dirs(bundles_dirs: List[str]) -> List[str]:
+    """Get a list of all bundles, including any overlays in all dirs passed.
+
+    :param bundles_dirs: the list of directories with bundles (hopefully).
+    :returns: List of filenames of bundles (yaml or yaml.j2)
+    """
+    return list(set(itertools.chain(
+        *(find_bundles(path) for path in bundles_dirs))))
+
+
+def add_channel_to(charms: List[str],
+                   bundle_filename: str,
+                   channel: str
+                   ) -> None:
+    """Add the candidate channel to the bundle as needed.
+
+    Adds a channel: candidate to the charms in the bundle that match one of the
+    charms in the :param:`charms` for ~openstack-charmers or
+    ~openstack-charmers-next.
+
+    :param charms: the list of charms that this will apply to.
+    :bundle_filename: the filename of the bundle to update.
+    """
+    new_file_name = "{}.new".format(bundle_filename)
+    with open(bundle_filename) as f:
+        file_lines = f.readlines()
+
+    new_lines = []
+
+    indent = None
+    for line in file_lines:
+        # line = line.rstrip()
+        if indent is not None:
+            if line.startswith(indent):
+                channel_match = CHANNEL_MATCH.match(line)
+                if channel_match:
+                    if channel_match[1] == indent:
+                        # replace the channel at the indent for the charm block
+                        new_lines.append("{}channel: {}\n".format(indent, channel))
+                        indent = None
+                        continue
+            else:
```

Comment:
I'm getting lost, I don't manage to validate the correctness of this function in my head, because:
1. there are different places where we add `channel:`
2. it depends on a variable indent that we reset at various places

Do you think you could add comments a bit in this function to guide me through? So for example what does it mean here that if the line doesn't start with the indent, then we add the channel? It's hard to prove in my head that this won't lead to adding `channel:` every second line. I trust you that you have run this function locally and it is working but it's not possible for me as a reviewer to prove its correctness in my head

Reply:
@AurelienLourot I've now added comments and a commentary to that part of the code; please let me know if it helps or hinders in understanding what the code is doing.  Thanks!

Reply:
Yes this is very clear now, thank you :pray: